packages/shopping-cart/src/state-based-actions.ts

```diff
@@ -0,0 +1,82 @@
+import debugFactory from 'debug';
+import { playQueuedActions } from './shopping-cart-reducer';
+import type {
+	ShoppingCartState,
+	ShoppingCartReducerDispatch,
+	LastValidResponseCart,
+	ActionPromises,
+	CacheStatus,
+	CartSyncManager,
+	SubscriptionManager,
+} from './types';
+
+const debug = debugFactory( 'shopping-cart:state-based-actions' );
+
+export function prepareFreshCartForInitialFetch(
+	state: ShoppingCartState,
+	dispatch: ShoppingCartReducerDispatch,
+	lastCacheStatus: CacheStatus | ''
+): void {
+	const { cacheStatus } = state;
+	if ( cacheStatus === 'fresh' && cacheStatus !== lastCacheStatus ) {
+		debug( 'triggering fetch of initial cart' );
+		dispatch( { type: 'FETCH_INITIAL_RESPONSE_CART' } );
+	}
+}
+
+function prepareInvalidCartForSync(
+	state: ShoppingCartState,
+	dispatch: ShoppingCartReducerDispatch,
+	lastCacheStatus: CacheStatus | ''
+): void {
+	const { queuedActions, cacheStatus } = state;
+	if (
+		queuedActions.length === 0 &&
+		cacheStatus === 'invalid' &&
+		cacheStatus !== lastCacheStatus
+	) {
+		debug( 'triggering sync of cart to server' );
+		dispatch( { type: 'REQUEST_UPDATED_RESPONSE_CART' } );
+	}
+}
+
+function isStatePendingUpdate( state: ShoppingCartState ) {
+	return state.queuedActions.length > 0 || state.cacheStatus !== 'valid';
+}
+
+export function createTakeActionsBasedOnState(
+	lastValidResponseCart: LastValidResponseCart,
+	actionPromises: ActionPromises,
+	syncManager: CartSyncManager,
+	subscriptionManager: SubscriptionManager
+): ( state: ShoppingCartState, dispatch: ShoppingCartReducerDispatch ) => void {
+	let lastCacheStatus: CacheStatus | '' = '';
+
+	const takeActionsBasedOnState = (
+		state: ShoppingCartState,
+		dispatch: ShoppingCartReducerDispatch
+	) => {
+		const { cacheStatus } = state;
+		debug( 'cache status before state-based-actions is', cacheStatus );
+		prepareFreshCartForInitialFetch( state, dispatch, lastCacheStatus );
+		prepareInvalidCartForSync( state, dispatch, lastCacheStatus );
+
+		syncManager.fetchInitialCartFromServer( state, dispatch );
+		syncManager.syncPendingCartToServer( state, dispatch );
+
+		if ( ! isStatePendingUpdate( state ) ) {
+			debug( 'updating lastValidResponseCart and resolving action promises' );
+			lastValidResponseCart.update( state.responseCart );
+			actionPromises.resolve( state.responseCart );
+		}
+
+		playQueuedActions( state, dispatch );
```

Comment:
I think there is a little race condition caused by the fact that `playQueuedActions` is deferred:
1. a sync response arrives, `RECEIVE_*` action sets status to `valid`
2. `playQueuedActions` is scheduled for later with `setTimeout`
3. before the timeout is called, an event loop tick squeezes in that dispatches an update action that changes status to `invalid`
4. only now the queued actions are dispatched, too. But the one in step 3 jumped the queue!

It couldn't happen when all dispatches were deferred, but now with synchronous dispatches it can.

One solution is that the "locked" reducer status is not determined solely by `cacheStatus === 'pending'`, but there is a standalone flag. It's set to `true` when setting `cacheStatus` to `pending`:
```js
case 'REQUEST_UPDATED_RESPONSE_CART':
  return {
    ...state,
    cacheStatus: 'pending',
    shouldQueue: true,
  };
```
but it's set back to `false` not when `cacheStatus` goes back to `valid`, but a bit later, in the deferred `playQueuedAction`, during the `CLEAR_QUEUED_ACTIONS` dispatch:
```js
case 'CLEAR_QUEUED_ACTIONS':
  return {
    ...state,
    queuedActions: [],
    shouldQueue: false
  };
```
As an optimization, we could merge all the queue info into just `queuedActions`:
- `null` value means "no queueing please"
- assigning `[]` means "please start queueing now"
- queueing actions means that the array gets new items if it exists

The `cacheStatusesForQueueing.includes` check then becomes `queuedActions != null`.

Reply:
I made `playQueuedActions` synchronous and that seems to have resolved any issues I was previously seeing. Do you think what you describe above is still an issue or can you think of any reason that it needs to be async? I can't think of a case where the remaining state-based actions (which are now just cart sync actions that eventually use `dispatch` anyway) would need to trigger a replay directly.

Reply:
I didn't realize that before, but yes, doing `playQueuedActions` synchronously is fine. It's no different from doing several batched synchronous `dispatch` calls, something we do all the time.

But one thing should be fixed: let's call `playQueuedActions` after `notifySubscribers`, not before. That way, subscribers have a chance to see the `state` as it was after the dispatched action and before the queued actions.

Reply:
🤔  Your suggestion made sense to me, but when I tried it, it broke loading checkout. It seems to be because of the nature of how actions are now dispatched. 

Previously, every dispatch ran and completed all its features before any other dispatch would run; they were atomic operations. Now that dispatches are synchronous, they are _recursive_, which creates a lot of complexity since a single operation could potentially do many things, including dispatches, before it completes. This is even more concerning now that I've discovered (in one of the other comments here) that React `setState` -> render is sometimes a synchronous process.

If we want to keep synchronous dispatches, I don't think there's much that we can do about that in general, but we may be able to reduce the number of dispatches being sent. I changed the action queuing process to run entirely as an enhancer to the reducer, so that the cart manager isn't aware of it and queued actions are replayed immediately after the cart becomes valid before the (enhanced) reducer even returns.

The downside of this is that if actions are queued, no subscribers will ever be able to see the cart state between when cart becomes valid and when the queued actions are dispatched, but in practice I don't think this will ever be a problem. The only time actions are queued is if we're waiting on a server response, and so there's really only one possible case of queued actions: We queue a change that will require another cart sync. In this case, the consumer is actually probably safer displaying is loading state for the combined two actions to prevent a flash, and in fact would have this behavior anyway because `isPendingUpdate` is always true if there are queued actions.

Reply:
> Your suggestion made sense to me, but when I tried it, it broke loading checkout.

But now the `playQueuedActions` is sychronous, there is one summary `notifyListeners` at the end, and yet loading checkout is not broken any more? Is that because there is only one `notifyListeners`, not multiple intermediate ones?

Reply:
I'm honestly not sure why it was broken before (although I still have the commits so we could find out). The issue appeared to be that by moving `playQueuedActions` after `notifySubscribers`, queued actions never ran at all if there were additional actions dispatched in the meantime? I was able to write a test that covers the issue here: https://github.com/Automattic/wp-calypso/blob/7c54b6d7330f04bfffc512779ac28e915c9ead58/packages/shopping-cart/test/cart-manager.ts#L148